Updates of copyright notices and license information in DebOps

drybjed opened the pull request:
This patch adds the SPDX license identifiers in files which were
detected by the 'reuse' tool and have existing copyright details.

I decided to review the copyright and licensing in DebOps and make it compliant with the [REUSE Specification](https://reuse.software/spec/). This hopefully will make the project easier to include in Debian archive in the future. The code is checked using the [`reuse`](https://pypi.org/project/reuse/) tool which eventually will be added to the Travis-CI test suite.

Review will take some time, but it's a one time process and after that it will be easy to stay compliant.

drybjed committed:
First pass of the copyright and license details

This patch adds the SPDX license identifiers in files which were
detected by the 'reuse' tool and have existing copyright details.

drybjed committed:
Second pass: copyright and license for internals

This patch adds the copyright and license information for various DebOps
monorepo internal codebase like CI tests, documentation infrastructure,
main documentation files and the like.

drybjed committed:
Add missing LICENSE files

The license files have been downloaded using the 'reuse download --all'
command, which downloaded them based on the information gathered via the
SPDX License Identifiers in the repository. The downloaded license files
will be included in the Asible Collections and Python packages of the
DebOps project.

SuperSandro2000 commented on LICENSES/MIT.txt:
```diff
@@ -0,0 +1,19 @@
+MIT License Copyright (c) <year> <copyright holders>
```
I have no clue what you could do with that line but leaving the default in seems strange.

drybjed replied:
As I, not a lawyer, understand it: the files in the `LICENSES/` directory are just for reference, because you MUST provide the license with the project. They shouldn't be changed, because quick check with `sha256sum` allows to verify if the license was changed or not

 Also, if I replaced the example text here with say, current year and my name and email, that probably would mean that suddenly DebOps is licensed under MIT license, right? Without changing the details here, it probably means that the MIT license applies only to the files that explicitly refer to it.

Also, the [answer](https://reuse.software/faq/#edit-license) from the REUSE FAQ about this:

> *Should I edit my license files?*

> You should never edit license files. When you use an existing license, you should always copy it verbatim.
> There are many reasons for why you should not alter license texts, but if you do alter the texts, you should use a different SPDX identifier for this license.

SuperSandro2000 replied:
I think the text you linked is about the legal text in the license which talks about your rights not the header. So lines 3 to 19 should stay the same. I think the best choice would be to drop the year and holders completely from this file.

Also sha sums are irrelevant cause as soon as you change any whitespace it is different which does not change the license what so ever.

drybjed replied:
I removed the `Copyright (c) <year> <copyright holders>` string from the file and the `reuse` script doesn't scream about any issues, so I suppose that's fine.

SuperSandro2000 commented on ansible/roles/owncloud/templates/srv/www/sites/themes/debops-template/defaults.php.j2:
```diff
@@ -13,7 +13,7 @@ Raw: https://raw.githubusercontent.com/owncloud/core/master/themes/example/defau
  * @author Björn Schießle <schiessle@owncloud.com>
  * @author Jan-Christoph Borchardt, http://jancborchardt.net
  * @copyright Copyright (c) 2015, ownCloud, Inc.
- * @license AGPL-3.0-only
```
I think this is some magic php comment and should maybe stay.

drybjed replied:
Sounds good, I'll bring it back then.

SuperSandro2000 commented on docs/_templates/breadcrumbs.html:
```diff
@@ -1,3 +1,7 @@
+{# Copyright (C) 2016      Robin Schneider <ypid@riseup.net> #}
```
Does this highlight mean anything or is it just some strange github thing?

drybjed replied:
It's just the syntax highliter being confused, probably it thinks these are some kind of CSS references.

drybjed committed:
[owncloud] Bring back the '@license' marker

drybjed committed:
[apt] Add copyright and license information

drybjed committed:
[ansible] Add copyright and license information

drybjed committed:
[ansible_plugins] Add copyright, license details

drybjed committed:
[docs] Correct issues with documentation includes

drybjed committed:
[apache] Add copyright and license information

drybjed committed:
[apt_cacher_ng] Add copyright, license information

drybjed committed:
[apt_install] Add copyright, license information

drybjed committed:
Remove empty "Copyright" string from MIT license

drybjed committed:
[apt_listchanges] Add copyright, license details

drybjed committed:
[apt_mark] Add copyright and license information

ypid reviewed (dismissed):
Looks like a great idea. I reviewed the PR and it looks good to me. I did not know reuse before. Thanks for maintaining!

drybjed committed:
[apt_preferences] Add copyright, license details

drybjed committed:
[apt_proxy] Add copyright and license information

drybjed committed:
[atd] Add copyright and license information

drybjed committed:
[docs] Fix wrong file include ranges

It seems that Sphinx counts the 'start-line' line numbers from 0, not
from 1.

drybjed committed:
[auth] Add copyright and license information

drybjed committed:
[authorized_keys] Add copyright, license details

drybjed committed:
[avahi] Add copyright and license details

drybjed committed:
[backup2l] Add copyright and license information

drybjed committed:
[boxbackup] Add copyright and license information

drybjed committed:
[console] Add copyright and license information

drybjed committed:
[core] Add copyright and license information

drybjed committed:
[cran] Add copyright and license information

drybjed committed:
[cron] Add copyright and license information

drybjed committed:
[cryptsetup] Add copyright and license information

drybjed committed:
[debops] Add copyright and license information

drybjed committed:
[apparmor] Add copyright and license information

drybjed committed:
[bitcoind] Add copyright and license information

drybjed committed:
[btrfs] Add copyright and license information

drybjed committed:
[dropbear_initramfs] Add copyright, license info

drybjed committed:
[firejail] Add copyright and license information

drybjed committed:
[foodsoft] Add copyright and license information

drybjed committed:
[fuse] Add copyright and license information

drybjed committed:
[homeassistant] Add copyright and license details

drybjed committed:
[kodi] Add copyright and license information

drybjed committed:
[snapshot_snapper] Add copyright and license info

drybjed committed:
[tor] Add copyright and license information

drybjed committed:
[volkszaehler] Add copyright and license details

drybjed committed:
[x2go_server] Add copyright and license details

drybjed committed:
Add copyright and license details to contrib plays

drybjed committed:
[debops_api] Add copyright and license information

drybjed committed:
[debops_fact] Add copyright and license details

drybjed committed:
[debops_legacy] Add copyright and license details

drybjed committed:
[dhcp_probe] Add copyright and license information

drybjed committed:
[dhcpd] Add copyright and license information

drybjed committed:
[dhparam] Add copyright and license information

drybjed committed:
[dnsmasq] Add copyright and license information

drybjed committed:
[docker_gen] Add copyright and license information

drybjed committed:
[docker_registry] Add copyright, license details

drybjed committed:
[docker_server] Add copyright, license information

drybjed committed:
[dokuwiki] Add copyright and license information

drybjed committed:
[dovecot] Add copyright and license information

drybjed committed:
[dpkg_cleanup] Add copyright and license details

drybjed committed:
[elastic_co] Add copyright and license information

drybjed committed:
[elasticsearch] Add copyright and license details

drybjed committed:
[environment] Add copyright and license details

drybjed committed:
[etc_aliases] Add copyright and license details

drybjed committed:
[etc_services] Add copyright and license details

drybjed committed:
[etckeeper] Add copyright and license information

drybjed committed:
[etesync] Add copyright and license information

drybjed committed:
[etherpad] Add copyright and license information

drybjed committed:
[fail2ban] Add copyright and license information

drybjed committed:
[fcgiwrap] Add copyright and license information

drybjed committed:
[docs] Fix wrong Sphinx keywords

drybjed committed:
[ferm] Add copyright and license information

drybjed committed:
[freeradius] Add copyright and license information

drybjed committed:
[gitlab] Add copyright and license information

drybjed committed:
[gitlab_runner] Add copyright, license details

drybjed committed:
[gitusers] Add copyright and license information

drybjed committed:
[golang] Add copyright and license information

drybjed committed:
[grub] Add copyright and license information

drybjed committed:
[gunicorn] Add copyright and license information

drybjed committed:
[hashicorp] Add copyright and license information

drybjed committed:
[hwraid] Add copyright and license information

drybjed committed:
[icinga] Add copyright and license information

drybjed committed:
[icinga_db] Add copyright and license information

drybjed committed:
[icinga_web] Add copyright and license information

drybjed committed:
[ifupdown] Add copyright and license information

drybjed committed:
[influxdata] Add copyright and license information

drybjed committed:
[ipxe] Add copyright and license information

drybjed committed:
[iscsi] Add copyright and license information

drybjed committed:
[java] Add copyright and license information

drybjed committed:
[journald] Add copyright and license information

drybjed committed:
[keyring] Add copyright and license information

drybjed committed:
[kibana] Add copyright and license information

drybjed committed:
[docs] Fix wrong import range

drybjed committed:
[kmod] Add copyright and license information

drybjed committed:
[ldap] Add copyright and license information

drybjed committed:
[librenms] Add copyright and license information

drybjed committed:
[libuser] Add copyright and license information

drybjed committed:
[libvirt] Add copyright and license information

drybjed committed:
[libvirtd] Add copyright and license information

drybjed committed:
[libvirtd_qemu] Add copyright, license information

drybjed committed:
[locales] Add copyright and license information

drybjed committed:
[logrotate] Add copyright and license information

drybjed committed:
[lvm] Add copyright and license information

drybjed committed:
[lxc] Add copyright and license information

drybjed committed:
[lxd] Add copyright and license information

drybjed committed:
[machine] Add copyright and license information

drybjed committed:
[mailman] Add copyright and license information

drybjed committed:
[mariadb] Add copyright and license information

drybjed committed:
[mariadb_server] Add copyright, license details

drybjed committed:
[mcli] Add copyright and license information

drybjed committed:
[memcached] Add copyright and license information

drybjed committed:
[minio] Add copyright and license information

drybjed committed:
[monit] Add copyright and license information

drybjed committed:
[mosquitto] Add copyright and license information

drybjed committed:
[mount] Add copyright and license information

drybjed committed:
[netbase] Add copyright and license information

drybjed committed:
[netbox] Add copyright and license information

drybjed committed:
[neurodebian] Add copyright and license details

drybjed committed:
[nfs] Add copyright and license information

drybjed committed:
[nfs_server] Add copyright and license information

drybjed committed:
[nginx] Add copyright and license information

drybjed committed:
[nodejs] Add copyright and license information

drybjed committed:
[nscd] Add copyright and license information

drybjed committed:
[nslcd] Add copyright and license information

drybjed committed:
[nsswitch] Add copyright and license information

drybjed committed:
[ntp] Add copyright and license information

drybjed committed:
[nullmailer] Add copyright and license information

drybjed committed:
[opendkim] Add copyright and license information

drybjed committed:
[owncloud] Add copyright and license information

drybjed committed:
[pam_access] Add copyright and license information

drybjed committed:
[persisten_paths] Add copyright, license details

drybjed committed:
[php] Add copyright and license information

drybjed committed:
[docs] Fix small mistakes in 'owncloud' role

drybjed committed:
[phpipam] Add copyright and license information

drybjed committed:
[phpmyadmin] Add copyright and license information

drybjed committed:
[pki] Add copyright and license information

drybjed committed:
[postconf] Add copyright and license information

drybjed committed:
[postfix] Add copyright and license information

drybjed committed:
[postgresql] Add copyright and license information

drybjed committed:
[postgresql_server] Add copyright, license details

drybjed committed:
[postldap] Add copyright and license information

drybjed committed:
[postscreen] Add copyright and license information

drybjed committed:
[postwhite] Add copyright and license information

drybjed committed:
[preseed] Add copyright and license information

drybjed committed:
[proc_hidepid] Add copyright, license information

drybjed committed:
[prosody] Add copyright and license information

drybjed committed:
[python] Add copyright and license information

drybjed committed:
[rabbitmq_management] Add copyright, license info

drybjed committed:
[rabbitmq_server] Add copyright, license details

drybjed committed:
[radvd] Add copyright and license information

drybjed committed:
[rails_deploy] Add copyright and license details

drybjed committed:
[redis_sentinel] Add copyright, license details

drybjed committed:
[redis_server] Add copyright, license information

drybjed committed:
[reprepro] Add copyright and license information

drybjed committed:
[resolvconf] Add copyright and license information

drybjed committed:
[resources] Add copyright and license information

drybjed committed:
[root_account] Add copyright, license information

drybjed committed:
[roundcube] Add copyright and license information

drybjed committed:
[rsnapshot] Add copyright and license information

drybjed committed:
[rstudio_server] Add copyright and license details

drybjed committed:
[rsyslog] Add copyright and license information

drybjed committed:
[ruby] Add copyright and license information

drybjed committed:
[salt] Add copyright and license information

drybjed committed:
[samba] Add copyright and license information

drybjed committed:
[saslauthd] Add copyright, license information

drybjed committed:
[secret] Add copyright and license information

drybjed committed:
[docs] Remove incorrect copyright attribution

This snippet was added with the 'owncloud' role copyright and license
notice by mistake.

drybjed committed:
[sks] Add copyright and license information

drybjed committed:
[slapd] Add copyright and license information

drybjed committed:
[smstools] Add copyright and license information

drybjed committed:
[snmpd] Add copyright and license information

drybjed committed:
[sshd] Add copyright and license information

drybjed committed:
[stunnel] Add copyright and license information

drybjed committed:
[sudo] Add copyright and license information

drybjed committed:
[swapfile] Add copyright and license information

drybjed committed:
[sysctl] Add copyright and license information

drybjed committed:
[sysfs] Add copyright and license information

drybjed committed:
[sysnews] Add copyright and license information

drybjed committed:
[system_groups] Add copyright, license information

drybjed committed:
[system_users] Add copyright and license details

drybjed committed:
[tcpwrappers] Add copyright and license details

drybjed committed:
[tftpd] Add copyright and license information

drybjed committed:
[tgt] Add copyright and license information

drybjed committed:
[tinc] Add copyright and license information

drybjed committed:
[tinyproxy] Add copyright and license information

drybjed committed:
[unattended_upgrades] Add copyright, license info

drybjed committed:
[unbound] Add copyright and license information

drybjed committed:
[users] Add copyright and license information

drybjed committed:
[wpcli] Add copyright and license information

drybjed committed:
[yadm] Add copyright and license information

drybjed committed:
Add copyright and license information in playbooks

ypid commented on .reuse/dep5:
```diff
@@ -0,0 +1,90 @@
+Copyright: 2015      Hartmut Goebel <h.goebel@crazy-compilers.com>
+Copyright: 2015-2019 Robin Schneider <ypid@riseup.net>
+Copyright: 2015-2019 DebOps <https://debops.org/>
+License: GPL-3.0-or-later
```
Not sure if GPL is the best license for OpenPGP keys. Maybe put it under Public Domain?

drybjed replied:
There is no "Public Domain" license in SPDX list [due to reasons](https://wiki.spdx.org/view/Legal_Team/Decisions/Dealing_with_Public_Domain_within_SPDX_Files). In my opinion these keys should be removed entirely from the repository - the `keyring` role can deal with GPG keys not published on the keyserver network by adding them to the local keyring directory. The role can mention that these GPG keys need to be downloaded separately. But since for now I wanted to capture the existing state which can be changed later, I went with the same license as the role itself.

I'm not sure what license should be applied to GPG keys... What do you think about [CC0](https://creativecommons.org/share-your-work/public-domain/cc0/) instead?

ypid replied:
CC0 is also fine. Actually, when I said Public Domain, I already included CC0 in my thought :)

drybjed replied:
OK, will do.

ypid commented:
Out of interest. How are you adding those headers? Is it somewhat tool-assisted based on the git history?

drybjed commented:
I want this PR to be a "demarcation point" between basically chaotic license and copyright information in DebOps in the past, and known good state in the future. So, no automation - all 3462 files are/were checked by hand by me and the copyright and license details written separately (I don't automate GPG signatures either, so yeah - all these commits were signed by me personally).

I wanted to capture the current state of the repository as-is, but at the same time I didn't want to do it for like 3-4 weeks combing through various `git` repositories of the project. Because of that mot all copyright information is accurate - to save time I looked through `git` history for most of the non-role related files in available repositories and I tried to capture rough years where a given author touched these files, but not all is there. For the role files, I copied the details from the `COPYRIGHT` files of each role treating all files under a given subdirectory with the same copyright and license data. In the future I hope that more people will help with filling up the missing copyright credits.

ypid commented:
Thanks for the details and your commitment! Sounds good.

drybjed committed:
Add copyright and license details in documentation

drybjed committed:
[docs] Change the license of included GPG keys

drybjed committed:
Add missing license texts in 'LICENSES/' directory

drybjed commented:
Final summary from the `reuse` tool:
```
# SUMMARY

* Bad licenses:
* Deprecated licenses:
* Licenses without file extension:
* Missing licenses:
* Unused licenses:
* Used licenses: AGPL-3.0-only, Apache-2.0, BSD-4-Clause, CC-BY-SA-4.0, CC0-1.0, GPL-2.0-only, GPL-2.0-or-later, GPL-3.0-or-later, MIT
* Read errors: 0
* Files with copyright information: 3462 / 3462
* Files with license information: 3462 / 3462

Congratulations! Your project is compliant with version 3.0 of the REUSE Specification :-)
```

drybjed committed:
Add Changelog entry about SPDX changes

drybjed committed:
[ci] Add 'reuse' tool to the Travis-CI test suite

drybjed committed:
[ci] Disable SPDX test for now to resolve conflict

ypid commented:
Well done!

Can you talk a bit about why you changed from GPL-3.0-only to GPL-3.0-or-later? Do you fear that it would be more difficult to switch to a new version later (as in the Linux Kernel for example)?

drybjed commented:
It's hard to tell if there ever will be a new version of the GPL license. The current one seems to cover most of the "loopholes", and the AGPL is an even more strict version for the SaaS world, I don't think that another iteration is necessary - but who knows... Perhaps the recent emergence of Commons Clause and similar efforts to fight with big players taking over projects and profiting fom them without giving back (something AGPL was probably meant to tackle) will result in another attempt at a new GPL license.

As for switching to the `-or-later` version within the DebOps project - I tried to make the status quo explicit... But now when I look at the license texts closely it seems that I made a mistake with giving all of the roles the `later` treatment. Since the scripts were done first, and they were licensed with `-or-later` version of the license, I must have assumed that the roles were licensed the same and I din't pay enough attention to the text of the license in the `COPYRIGHT`files which don't explicitly state this.

Hmmm, what to do what to do... I'm not really opposed at setting the `GPL-3.0-or-later` version to be the default. if we assume that a potential new version of the GPL license will be an improvement over the current 3.0 one, having an option to switch to it might be useful in the future. This also aligns in my opinion with DebOps being a long term project - future-proofing it might be handy.

ypid commented:
My personal (and at work) policy is to go with `3.0-only` because I would like to review new versions of the GPL first. But I am not sure how well this scales thus I am open to have `3.0-or-later` in DebOps. The issue is that such license upgrade (GPL-2.0-only to GPL-3.0) might require approval from all copyright owners. This is impossible for projects like the Linux Kernel.

Nextcloud is `AGPL-3.0-or-later` (ref: https://github.com/nextcloud/server/blob/master/package.json) and I trust them with making a responsible choice as they had a bit of a learning curve :)

drybjed commented:
I see. Do you think that the current state with `GPL-3.0-or-later` everywhere is fine, or should I go over the files again, revert to `GPL-3.0-only` where `-or-later` is not explicitly permitted, after that add missing copyright holders to files... And then we can switch the license over when all contributors to a given section of the code are fine with it? I think that as long as the source is not releases as tarballs/Ansible Collections/Python packages, we have a "window of opportunity" to straighten out any SPDX identifiers that were wrongly selected, as long as the changes are documented in the commit history.

Also, any thoughts about adopting [Developer Certificate of Origin](https://developercertificate.org/) for DebOps? That would entail [adding `Signed-Off-By:` string](https://stackoverflow.com/a/1962112) in the commit text. This can be done using `git commit -s` command, but would that add additional hurdle for contributing to DebOps? Or is there no benefit at this point?

ypid commented:
I fully agree with your assesment of the situation. It only counts when it is included in a release. The safer thing to do would be to revert to `GPL-3.0-only`.

> Developer Certificate of Origin

I will have to check that. I have not looked into this before.

drybjed commented:
Very well then, round two it is.
